crates/lib3h/src/transport/transport_multiplex/mplex.rs

```diff
@@ -0,0 +1,377 @@
+use crate::transport::{error::*, protocol::*};
+use detach::prelude::*;
+use lib3h_ghost_actor::prelude::*;
+use std::{any::Any, collections::HashMap};
+use url::Url;
+
+use super::{LocalRouteSpec, TransportMultiplexRoute};
+
+enum MplexToInnerContext {
+    AwaitBind(
+        GhostMessage<RequestToChild, RequestToParent, RequestToChildResponse, TransportError>,
+    ),
+    AwaitSend(
+        GhostMessage<RequestToChild, RequestToParent, RequestToChildResponse, TransportError>,
+    ),
+}
+
+type SelfContextEndpoint<Context> = GhostContextEndpoint<
+    Context,
+    RequestToParent,
+    RequestToParentResponse,
+    RequestToChild,
+    RequestToChildResponse,
+    TransportError,
+>;
+
+pub struct TransportMultiplex {
+    // our parent channel endpoint
+    endpoint_parent: Option<TransportActorParentEndpoint>,
+    // our self channel endpoint
+    endpoint_self: Detach<SelfContextEndpoint<()>>,
+    // ref to our inner transport
+    inner_transport: Detach<TransportActorParentWrapperDyn<MplexToInnerContext>>,
+    // our map of endpoints connecting us to our Routes
+    route_endpoints: Detach<HashMap<LocalRouteSpec, SelfContextEndpoint<()>>>,
+}
+
+impl TransportMultiplex {
+    /// create a new TransportMultiplex Instance
+    pub fn new(inner_transport: DynTransportActor) -> Self {
+        let (endpoint_parent, endpoint_self) = create_ghost_channel();
+        let endpoint_parent = Some(endpoint_parent);
+        let endpoint_self = Detach::new(endpoint_self.as_context_endpoint("mplex_to_parent_"));
+        let inner_transport = Detach::new(GhostParentWrapperDyn::new(
+            inner_transport,
+            "mplex_to_inner_",
+        ));
+        Self {
+            endpoint_parent,
+            endpoint_self,
+            inner_transport,
+            route_endpoints: Detach::new(HashMap::new()),
+        }
+    }
+
+    /// create a route for a specific agent_id + space_address combination
+    /// we are wrapping a network/machine-level gateway with a machine
+    /// space_address and machineId... the space_address + agent_id parameters
+    /// for this function are the higher-level notions for the AgentSpaceGateway
+    pub fn get_agent_space_route(
+        &mut self,
+        space_address: String,
```

Comment:
Why is this a raw string? Can it instead be a `lib3h_protocol::Address` or `hcid`? Same w/agent id.

Reply:
done